packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx

```diff
@@ -153,32 +153,47 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
   const toggleButtonRef = React.useRef<HTMLSpanElement>(null);
   const mounted = React.useRef(false);
 
+  // Record to store all extra information (e.g. original indexes), where the key is the id of the option
+  const optionsExtraInfoRef = React.useRef<
+    Record<string, { __originalIndex: number }>
+  >({});
+
+  // Function to initialize extra info
+  const initializeExtraInfo = React.useCallback(() => {
+    optionsRef.current.forEach((option, index) => {
+      optionsExtraInfoRef.current[getOptionId(option, id)] = {
+        __originalIndex: index,
+      };
+    });
+  }, [id]);
+
+  // Latest value of the valueProp prop
+  const valuePropRef = React.useRef(valueProp);
+  valuePropRef.current = valueProp;
+
   // Latest value of the onChange prop
   const onChangeProp = React.useRef(onChange);
   React.useEffect(() => {
     onChangeProp.current = onChange;
   }, [onChange]);
```

Comment:
I like the changes you have done with the options and vaue refs, and it made me realize we don't need this useEffect either. We _always_ want the latest value.

```suggestion
  onChangeProp.current = onChange;
```

Reply:
Added new private hook to save latest value.